src/pybind/mgr/cephadm/services/ha_rgw.py

```diff
@@ -0,0 +1,161 @@
+                    code, '\n'.join(err)))
+
+        ret, keyring, err = self.mgr.check_mon_command({
+            'prefix': 'auth get-or-create',
+            'entity': self.get_auth_entity(daemon_id),
+            'caps': [],
+        })
+
+        logger.info('Create daemon %s on host %s with spec %s' % (
+            daemon_id, host, spec))
+        return daemon_spec
+
+    def generate_config(self, daemon_spec: CephadmDaemonSpec) -> Tuple[Dict[str, Any], List[str]]:
+
+        daemon_id = daemon_spec.daemon_id
+        host = daemon_spec.host
+
+        service_name: str = "HA_RGW." + daemon_id.split('.')[0]
+        if service_name in self.mgr.spec_store.specs:
+            spec = cast(HA_RGWSpec, self.mgr.spec_store.specs[service_name])
+
+            rgw_daemons = self.mgr.cache.get_daemons_by_type('rgw')
+            rgw_servers = []
+            for daemon in rgw_daemons:
+                rgw_servers.append(self.rgw_server(
+                    daemon.name(), resolve_ip(daemon.hostname)))
+
+            # virtual ip address cannot have netmask attached when passed to haproxy config
+            # since the port is added to the end and something like 123.123.123.10/24:8080 is invalid
+            virtual_ip_address = spec.virtual_ip_address
+            if "/" in str(spec.virtual_ip_address):
+                just_ip = str(spec.virtual_ip_address).split('/')[0]
+                virtual_ip_address = just_ip
+
+            ha_context = {'spec': spec, 'rgw_servers': rgw_servers,
+                          'virtual_ip_address': virtual_ip_address}
+
+            haproxy_conf = self.mgr.template.render('services/haproxy/haproxy.cfg.j2', ha_context)
+
+            config_file = {
+                'files': {
+                    "haproxy.cfg": haproxy_conf,
+                }
+            }
+
+            return config_file, []
+        else:
+            config_file = {'files': {}}
+            return config_file, []
+
+
+class KeepAlivedService(CephService):
+    TYPE = 'keepalived'
+
+    def prepare_create(self, daemon_spec: CephadmDaemonSpec[HA_RGWSpec]) -> CephadmDaemonSpec:
+        assert self.TYPE == daemon_spec.daemon_type
+        assert daemon_spec.spec
+
+        daemon_id = daemon_spec.daemon_id
+        host = daemon_spec.host
+        spec = daemon_spec.spec
+
+        out, err, code = self.mgr._run_cephadm(host, cephadmNoImage, 'verify-prereqs',
+                                               ['--service-type', 'keepalived'],
+                                               no_fsid=True, error_ok=True)
+
+        if code:
+            raise OrchestratorError(
+                'Keepalived failed prereqs check with code: %d, stderr:%s' % (
+                    code, '\n'.join(err)))
+
+        ret, keyring, err = self.mgr.check_mon_command({
+            'prefix': 'auth get-or-create',
+            'entity': self.get_auth_entity(daemon_id),
+            'caps': [],
+        })
```

Comment:
why does keepalived need an auth entity?

Reply:
You're right. Not needed at all